### PR TITLE
LANG-740 - Implementation of a Memomizer

### PR DESCRIPTION
**Pull request by jamessawle**
Enclosed is an implementation of the Memoizer class, along with some unit tests through the main flow of the code.

Is a copy of https://github.com/apache/commons-lang/pull/63 against the new 'master' branch

### CODE REVIEW

**Comment by britter** [comment]
Hello @jamessawle,

thank you for this PR. It looks very good to me. However there are a few this we need to adjust before I can apply it. Please have a look at my inline comments.

Thank you!

File: src/main/java/org/apache/commons/lang3/concurrent/Computable.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * <p>Definition of an interface for a wrapper around a calculation that takes a single parameter and returns a result.</p>
+ * <p/>
+ * <p>This interface allows for wrapping a calculation into a class so that it maybe passed around an application.</p>
+ *
+ * @param <A> the type of the input to the calculation
+ * @param <V> the type of the output of the calculation
+ */
+public interface Computable<A, V> {
+
+	/**
```
Comment (britter) [inline src/main/java/org/apache/commons/lang3/concurrent/Computable.java]: Please only use spaces for indentation

File: src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		replay(computable);
+
+		assertEquals("Should call computable first time", input, memoizer.compute(input));
+		assertEquals("Should not call the computable the second time", input, memoizer.compute(input));
+	}
+
+	@Test(expected = IllegalStateException.class)
+	public void testDefaultBehaviourNotToRecalculateExecutionExceptions() throws Exception {
+		Integer input = 1;
+		Integer answer = 3;
```
Comment (britter) [inline src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java]: answer is unused

File: src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+			//Should always be thrown the first time
+		}
+
+		memoizer.compute(input);
+	}
+
+	@Test(expected = IllegalStateException.class)
+	public void testDoesNotRecalculateWhenSetToFalse() throws Exception {
+		Integer input = 1;
+		Integer answer = 3;
```
Comment (britter) [inline src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java]: answer is unused

File: src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+package org.apache.commons.lang3.concurrent;
```
Comment (britter) [inline src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java]: Apache license header is missing

File: src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static org.easymock.EasyMock.expect;
+import static org.easymock.EasyMock.replay;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+@RunWith(EasyMockRunner.class)
+public class MemoizerTest {
```
Comment (britter) [inline src/test/java/org/apache/commons/lang3/concurrent/MemoizerTest.java]: Wouldn't it make sense to create a test that uses an executor service to let several threads in parallel access the same Memoizer instance?

File: src/main/java/org/apache/commons/lang3/concurrent/Memoizer.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	 * <p>This cache will also cache exceptions that occur during the computation if the {@code recalculate} parameter is
+	 * the constructor was set to {@code false}, or not set. Otherwise, if an exception happened on the previous
+	 * calculation, the method will attempt again to generate a value.</p>
+	 *
+	 * @param arg the argument for the calculation
+	 * @return the result of the calculation
+	 * @throws InterruptedException  thrown if the calculation is interrupted
+	 * @throws IllegalStateException a wrapper around any checked exception that occurs during the computation of the result
+	 */
+	public V compute(final A arg) throws InterruptedException, IllegalStateException {
```
Comment (peteryhwong) [inline src/main/java/org/apache/commons/lang3/concurrent/Memoizer.java]: Please use Overrides annotation. Moreover, IlegalStateException is a runtime (unchecked) exception and so need not be declared at the method signature.

**Comment by PascalSchumacher** [comment]
Thanks!

**Pull request closed without merging**